### PR TITLE
refactor: only add / remove minimul number of cdroms

### PR DESCRIPTION
**Pull request by Hi-Angel**
In PR #344 a feature to reuse an existing VM is being added. As part of that feature I had to factor out from `AddCdrom` the logic of creating a CD-rom as opposed to just mounting an image. Recently merged feature `reattach_cdroms` turned out to be a new user of the older logic, which made rebasing #344 non-trivial.

In this PR I refactor the `reattach_cdroms` logic to avoid unnecessarily calling `EjectCdroms`, `RemoveCdroms` and `AddCdrom` by calculating amount of CDroms that we want to have added/removed, and only calling the mentioned functions for that amount *(see the last commit)*.

This is a useful change in itself, disregarding whether #344 exists, so I'm sending it separately from the other PR. And it should simplify rebasing #344 later.

### CODE REVIEW

**Comment by nywilken** [comment]
@Hi-Angel thanks for opening up this quick refactor PR. I was planning to release the vSphere plugin with the latest changes. 

Should I wait until this is merged? Or is the refactor for supporting #344

**Comment by Hi-Angel** [comment]
I don't have any preference here. On one hand, I presume the code that is already in the repo has probably seen more testing than the new one, so regressions are less likely. On the other hand, this refactor reduces a bit calls which *(per my understanding)* communicate to ESXi via network, which is slow. So if you decide to wait for it to land, the new functional might be a tiny bit faster.

**Comment by nywilken** [comment]
> I don't have any preference here. On one hand, I presume the code that is already in the repo has probably seen more testing than the new one, so regressions are less likely. On the other hand, this refactor reduces a bit calls which _(per my understanding)_ communicate to ESXi via network, which is slow. So if you decide to wait for it to land, the new functional might be a tiny bit faster.

@Hi-Angel that's a good call out on the testing part. We release every week, or shoot for it when we have pending changes. I think we can focus on getting this PR reviewed and merged for the next release to give the new feature time to setting.

**Comment by Hi-Angel** [comment]
Any news here?

**Comment by tenthirtyam** [comment]
> Any news here?

I've rebased from `main` and have squashed the commits. I'll review this w/in the next week.

**Comment by Hi-Angel** [comment]
> > Any news here?
> 
> I've rebased from `main` and have squashed the commits. I'll review this w/in the next week.

Why have you squashed the commits before review? 😅 I mean, I know the project squashes them on merge, but don't you want for review purposes to see each functional change separated?

**Comment by tenthirtyam** [comment]
It's all good - I've read through them already.

File: builder/vsphere/driver/vm_cdrom.go
Comment (nywilken) [inline builder/vsphere/driver/vm_cdrom.go:71]: I don't understand why this check. How is this removal different from the removal happening above. I see there are different devices types but I don't understand why it only happens if cdroms is greater than nCdroms which will only occur if nCdroms is 0. Is that right?

File: builder/vsphere/driver/vm_cdrom.go
Comment (nywilken) [inline builder/vsphere/driver/vm_cdrom.go:58]: Working through this I see that if nCdroms is greater than 0, then nCdroms will be removed. You do this by subslicing the cdroms slice which will remove the devices from 0 up until nCdroms-1. But if you nCdroms is 0 all devices get removed. That is not the behavior I would expect. 

Instead of changing the function signature for RemoveCdroms I suggest creating a new method called RemoveVirtualCdromsN(int)

```
//RemoveVirtualCdromsN removes up to n CDRoms from the associated vm.
//An error will occur If n is larger then the attached CDRoms. 
//n == 0 results in no CDRoms being removed
func (vm *VirtualMachineDriver) RemoveVirtualCdromsN(n int) error {
        if n == 0 {
		return nil
	}	
        devices, err := vm.Devices()
	if err != nil {
		return err
	}
	cdroms := devices.SelectByType((*types.VirtualCdrom)(nil))
        if (n < 0) || (n > len(cdroms)){
         return errors.New("invalid number: n must be between 0 and %d, len(cdroms))
        }
       
        cdroms = cdroms[:n]
	if err = vm.RemoveDevice(true, cdroms...); err != nil {
		return err
	}

//// I don know what needs to happen with the 
// sata := devices.SelectByType((*types.VirtualAHCIController)(nil)) ....
	return nil
}

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_reattach_cdrom.go:51]: But this way you can't have the type-hint. I mean, I can change that if you want… it's just that while working with Go code, the guessing game of "which type does this variable have?" is among the worst problems. It's mostly due lang design, but still I'm trying my best put type-hints when possible, so other people would have simpler time debugging the code.

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_reattach_cdrom.go:52]: Correct. To clarify:

>  we want to reattach 2 cdroms but have 4 attached len(s.CDRomConfig.ISOPaths) this logic will only remove 2 since they will no longer be needed

…and the remaining two will have cdroms ejected 

> If we have 4 attached and want reattach 4 we remove none to essentially reuse attached CDROMs

…and all 4 of them have cdroms ejected

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_reattach_cdrom.go:65]: Will do

File: builder/vsphere/driver/vm_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/driver/vm_cdrom.go:71]: Hmmm… I think that might be a mistake… I'll take a closer look tomorrow

File: builder/vsphere/driver/vm_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/driver/vm_cdrom.go:58]: Sure, will do

File: builder/vsphere/driver/vm_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/driver/vm_cdrom.go:71]: Ah, I got it. What this `if` does is it basically checks if we remove all cdroms or not *(that is due to `nCdroms` having special value `0`)*.

Yeah, that's confusing. Anyway, I'll follow with your suggestion below of having a separate function for this, that should resolve the readability of this code as well.

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_reattach_cdrom.go]: Done.

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_reattach_cdrom.go]: Done.

File: builder/vsphere/driver/vm_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/driver/vm_cdrom.go]: The code was removed

File: builder/vsphere/driver/vm_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/driver/vm_cdrom.go]: Done. I noticed though that we already have `RemoveCdroms` without infix `Virtual`, which makes sense because typically the plugin is dealing with "virtual hardware", not a passthrough. So I took the liberty of naming it a `RemoveNCdroms`, i.e. without the `Virtual` infix.

Tell if you want that to be changed.

**Comment by Hi-Angel** [comment]
Everything seems addressed

**Review by nywilken — APPROVED** [review]
@Hi-Angel then updated changes look good to me. After your explanation dropping `Virtual` from the function name makes senses. Thanks for responding so quickly to the review. 

@tenthirtyam would you like to give this another set of eyes before I merge?

File: builder/vsphere/driver/vm_cdrom.go
Comment (nywilken) [reply builder/vsphere/driver/vm_cdrom.go]: Thanks for the explanation.

**Comment by tenthirtyam** [comment]
> @tenthirtyam would you like to give this another set of eyes before I merge?

Will review and test as well and update you.

File: builder/vsphere/common/step_add_cdrom.go
Comment (tenthirtyam) [inline builder/vsphere/common/step_add_cdrom.go:87]: If calling out a bug (or is it actually a limitation) in `vmware/govmomi`, it would be ideal to include a reference so that this can be reviewed in the future.

File: builder/vsphere/common/step_add_cdrom.go
Comment (tenthirtyam) [inline builder/vsphere/common/step_add_cdrom.go:90]:
```suggestion
			state.Put("error", fmt.Errorf("invalid path: empty string"))
```

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (tenthirtyam) [inline builder/vsphere/common/step_reattach_cdrom.go:51]: `nAttachableCdroms`

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (tenthirtyam) [inline builder/vsphere/common/step_reattach_cdrom.go:52]: `nAttachableCdroms`

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (tenthirtyam) [inline builder/vsphere/common/step_reattach_cdrom.go:66]:
```suggestion
		// Eject media from the existing CD-ROM devices.
```

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (tenthirtyam) [inline builder/vsphere/common/step_reattach_cdrom.go:75]: `nAttachableCdroms`

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (tenthirtyam) [inline builder/vsphere/common/step_reattach_cdrom.go:88]: `nAttachableCdroms`

File: builder/vsphere/common/step_reattach_cdrom_test.go
Comment (tenthirtyam) [inline builder/vsphere/common/step_reattach_cdrom_test.go:30]:
```suggestion
			name: "Successfully attach 3 additional CD-ROM devices",
```

File: builder/vsphere/common/step_reattach_cdrom_test.go
Comment (tenthirtyam) [inline builder/vsphere/common/step_reattach_cdrom_test.go:58]:
```suggestion
			name: "Successfully attach 2 fewer CD-ROM devices than initially added.",
```

File: builder/vsphere/common/step_reattach_cdrom_test.go
Comment (tenthirtyam) [inline builder/vsphere/common/step_reattach_cdrom_test.go:86]:
```suggestion
			name: "Successfully attach the same number of CD-ROMs as initially added.",
```

File: builder/vsphere/driver/vm_cdrom.go
Comment (tenthirtyam) [inline builder/vsphere/driver/vm_cdrom.go:78]:
```suggestion
// RemoveNCdroms removes up to n CD-ROMs from the image.
// An error will occur If n is larger then the attached CD-ROM count.
// n == 0 results in no CD-ROMs being removed.
```

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
Minor changes requested.

File: builder/vsphere/common/step_add_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_add_cdrom.go:87]: Well… I currently don't have resources for creating a minimal testcase for this, so let's just rename this to a "limitation" if that's okay with you

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_reattach_cdrom.go]: Hmm, well, I suppose I can put it before `nAttachableCdroms` calculation… I initially removed it because the whole point of the function is to process `reattach_cdroms` option. I.e. if you are inside this function, then you already know what this code supposed to do.

But I suppose the comment wouldn't hurt either, so I put it back.

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_reattach_cdrom.go]: Done!

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_reattach_cdrom.go]: Done!

File: builder/vsphere/common/step_add_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_add_cdrom.go]: Done!

File: builder/vsphere/common/step_add_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_add_cdrom.go]: Renamed per comment above

File: builder/vsphere/driver/vm_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/driver/vm_cdrom.go]: Done!

File: builder/vsphere/common/step_reattach_cdrom_test.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_reattach_cdrom_test.go]: Done! I removed a dot at the end because from looking at other tests I see the sentence ends with no dot at the end. Please tell if you want me to add it.

File: builder/vsphere/common/step_reattach_cdrom_test.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_reattach_cdrom_test.go]: Done! I removed a dot at the end because from looking at other tests I see the sentence ends with no dot at the end. Please tell if you want me to add it.

File: builder/vsphere/common/step_reattach_cdrom_test.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_reattach_cdrom_test.go]: Done!

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_reattach_cdrom.go]: Done!

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_reattach_cdrom.go]: Done!

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_reattach_cdrom.go]: Done!

File: builder/vsphere/common/step_reattach_cdrom.go
Comment (Hi-Angel) [reply builder/vsphere/common/step_reattach_cdrom.go]: Done!

File: builder/vsphere/driver/vm_cdrom.go
Comment (tenthirtyam) [reply builder/vsphere/driver/vm_cdrom.go]: Resolved in 860f7101405d68474e1acadf652ef640f3e43b2b.

File: builder/vsphere/common/step_reattach_cdrom_test.go
Comment (tenthirtyam) [reply builder/vsphere/common/step_reattach_cdrom_test.go]: Resolved in 860f7101405d68474e1acadf652ef640f3e43b2b.

File: builder/vsphere/common/step_reattach_cdrom_test.go
Comment (tenthirtyam) [reply builder/vsphere/common/step_reattach_cdrom_test.go]: Resolved in 860f7101405d68474e1acadf652ef640f3e43b2b.

File: builder/vsphere/common/step_reattach_cdrom_test.go
Comment (tenthirtyam) [reply builder/vsphere/common/step_reattach_cdrom_test.go]: Resolved in 860f7101405d68474e1acadf652ef640f3e43b2b.

**Pull request merged**
